### ISSUE TITLE
Proposal: `AsyncFunction`

### ISSUE THREAD

**Issue by sindresorhus** [sindresorhus/type-fest#121]
To make it easier to constrain a type to be an async function. We can currently do `T extends Function`, but there's no way to enforce it to be async.

We already have an internal type for this:

https://github.com/sindresorhus/type-fest/blob/78019f42ea888b0cdceb41a4a78163868de57555/source/async-return-type.d.ts#L3

But I'm thinking it could be more reusable:

```ts
type AsyncFunction<Arguments extends readonly any[] = readonly any[], ReturnValue = unknown> = (...arguments_: Arguments) => Promise<ReturnValue>;
```

Thoughts?

**Comment by papb** [comment]
I am neither in favor nor against it :sweat_smile: I'd use it only if I already had `type-fest` as a dependency. It is a type very easy to write, I don't mind having to declare `type AsyncFunction = (...args: any[]) => Promise<unknown>` whenever I need it.

As for the extra reusability you are proposing, to me it seems easier to just type what I want directly instead.

**Comment by sindresorhus** [comment]
> I'd use it only if I already had type-fest as a dependency.

That applies to many of the existing types. The point of `type-fest` is not only to be a dependency, but also be easily copy-pastable.

> As for the extra reusability you are proposing, to me it seems easier to just type what I want directly instead.

You might need it for a generic function.

**Comment by papb** [comment]
> That applies to many of the existing types.

Fair enough :grinning: I guess I am in favor of it then :)

**Comment by fregante** [comment]
Similarly, can this be a regular function as well? I lost count of how many times I had to write the latter:

```js
type AsyncFunction = (...args: any[]) => Promise<unknown>; 
type AnyFunction = (...args: any[]) => unknown; 
```

**Comment by sindresorhus** [comment]
Should probably be `AnyAsyncFunction` and `AnyFunction` then for consistency.

**Comment by papb** [comment]
"Fun" fact: `AnyFunction` should actually be `(...args: any[]) => any` instead of `(...args: any[]) => unknown` because `(a: never) => void` only extends the former.

**Comment by fregante** [comment]
All TypeScript facts are "fun". _In fact_ it should called "Fun"Script

**Comment by papb** [comment]
By the way, what is the difference between `(...args: any[]) => any` and `Function`?

**Comment by fregante** [comment]
> `(...args: any[]) => any` and `Function`

Not sure but last time I tried using `Function` it wouldn't accept it (I think I was trying to use `<T extends Function>` and it errored with `yourFunction is not compatible with Function`)

**Comment by fregante** [comment]
Here, a shortened version of what I'm working on:

```ts
function memoizeInflight<TFunction extends Function>(function_: TFunction): TFunction {
	return (async (...arguments_) => {
		return function_(...arguments_);
	}) as TFunction;
}

```
> line 1, eslint:
> Don't use `Function` as a type. Use a specific function type instead, like `() => void`. (@typescript-eslint/ban-types)XO(@typescript-eslint/ban-types)

> line 2, TypeScript:
> Conversion of type '(...arguments_: any[]) => Promise<any>' to type 'TFunction' may be a mistake because neither type sufficiently overlaps with the other. If this was intentional, convert the expression to 'unknown' first.
>   '(...arguments_: any[]) => Promise<any>' is assignable to the constraint of type 'TFunction', but 'TFunction' could be instantiated with a different subtype of constraint 'Function'.ts(2352)

Replace `Function` with `(...args: any[]) => unknown` and everything passes

**Comment by papb** [comment]
@fregante Maybe you shortened the example too much, but what if I pass some synchronous function to `memoizeInflight`, then TypeScript will tell me I have the exact same type, but at runtime you're actually giving me an async function, which clearly is not the same, so I'd say the fact that TS was complaining before is a good thing...

**Comment by papb** [comment]
Hmm however I just found that `Function` is not assignable to `(...args: any[]) => any`. [Just asked on SO](https://stackoverflow.com/q/64380905/4135063).

**Comment by sindresorhus** [comment]
Let's go with 3 types:

```ts
type AsyncFunction<Arguments extends readonly any[] = readonly any[], ReturnValue = unknown | void> = (...arguments_: Arguments) => Promise<ReturnValue>;

type AnyAsyncFunction = (...arguments_: readonly any[]) => Promise<unknown | void>;

type AnyFunction = (...arguments_: readonly any[]) => unknown | void;
```

I realized that any-type functions should not have generic arguments as they are type-erasers, but I think a generic `AsyncFunction` type could be useful for readability.

**Comment by Richienb** [comment]
@sindresorhus Doesn't this mean:

```ts
type AnyAsyncFunction = AsyncFunction;
```

**Comment by sindresorhus** [comment]
@Richienb Yes, but I think it's still useful to define `AnyAsyncFunction` explicitly as it will be slightly faster / easier for TS to handle.

**Comment by Richienb** [comment]
> it will be slightly faster / easier for TS to handle.

How much faster? This is an opportunity to save writing code and to make it easier to maintain.

**Comment by sindresorhus** [comment]
> "Fun" fact: AnyFunction should actually be (...args: any[]) => any instead of (...args: any[]) => unknown because (a: never) => void only extends the former.

@papb Any source for that?

And it seems like an edge-case to have an argument as `never` type.

The normal cases work fine, so I don't think it's worth caring about: https://www.typescriptlang.org/play?#code/C4TwDgpgBAggdiAYgVzgY2ASwPZygXigAoA6MgQwCcBzZAWwjmAGcB9ALinIQG0BdAJQEAfFFQBrONgDucANwAoBaEiwAjAWJEh+UQDdsmACZCIAD2CMjzWAhTosuKAH4oGzgAYFAem+63SirQMABMmkRE5JzMwJSYcNQ6+oYmUOaWcNa2SKgYOHiu7lBevv5qQA

**Comment by sindresorhus** [comment]
> How much faster? This is an opportunity to save writing code and to make it easier to maintain.

Probably negligible. I still find it more readable to have them separate since it's so little extra code and it also makes it easier to copy-paste the individual types. Let see what others think.

**Comment by Richienb** [comment]
Since these aren't utilities, do they fit as a "basic" type?

**Comment by sindresorhus** [comment]
Yes

**Comment by papb** [comment]
> > "Fun" fact: AnyFunction should actually be (...args: any[]) => any instead of (...args: any[]) => unknown because (a: never) => void only extends the former.
> 
> @papb Any source for that?

I don't remember where/how I found this, and I have no idea on the reasoning for this TS behavior (if any), but it's true, you can see in [this Playground](https://www.typescriptlang.org/play?#code/C4TwDgpgBAggdiAYgVzgY2ASwPZygXigAoA6MgQwCcBzZAWwjmAGcB9ALinIQG0BdAJQEAfFFQBrONgDucANwAoBaEiwAjAWJEh+UQDdsmACZCIAD2CMjzWAhTosuKAH4oGzgAZFAem+63SirQMABMmkRE5JzMwJSYcNQ6+oYmUOaWcNa2SKgYOHiu7lBeCr7+aoHg0ADqEJiURvZ5ToSRnHAQehCUSVAGxorKVbAAzJq19Y25jnjpVjbwOQ75Lm5Qnj5+oh6VqgBCEMCWlItNM+FkJFS0DExsnNwg-L2Pg0GwACzjdQ1nK3OZGwHI7dU7TFaFdbFTblBRAA)

> And it seems like an edge-case to have an argument as never type.

I think this edge-case can happen in some situations involving generic types:

```ts
type Foo<T> = (arg: T) => void;
```

If I have some type like that, I would want `Foo<T> extends AnyFunction` to always be true. And someone might pass `never` to `T` at some point, especially if writing another complex conditional type.

**Comment by sindresorhus** [comment]
> Also, I don't see any downsides on using => any over => unknown here, do you?

I try very hard to not use `any` whenever possible as it's less strict. In this case, `any` is fine though.
